frontends/PyCDE/src/pycde/types.py

```diff
@@ -42,17 +44,35 @@ def maybe_create_alias(self, inner_type, name):
       if name in self.registered_aliases:
         if alias != self.registered_aliases[name]:
           raise RuntimeError(
-              f"Re-defining type alias for {name}! "\
-              f"Given: {inner_type}, "\
-              f"existing: {self.registered_aliases[name].inner_type}"
-          )
+              f"Re-defining type alias for {name}! "
+              f"Given: {inner_type}, "
+              f"existing: {self.registered_aliases[name].inner_type}")
         return self.registered_aliases[name]
 
       self.registered_aliases[name] = alias
       return alias
 
     return inner_type
 
+  def declare_types(self, mod):
+    if not self.registered_aliases:
+      return
+
+    if not self.type_scopes.get(mod):
+      with mlir.ir.InsertionPoint.at_block_begin(mod.body):
+        self.type_scopes[mod] = hw.TypeScopeOp.create(self.TYPE_SCOPE)
+        self.declared_aliases[mod] = set([])
+
+    # TODO: at this point, we should ensure that any aliases referring to other
```

Comment:
I'm leaving this to future work in this PR, but FYI I will begin that work now, since I think it's pretty important to get right sooner rather than later.

Reply:
Regarding declaring types before other types that may use them, I thought about how these dependencies can arise, and it occurred to me that end users must declare the depended-on types through this module before they can declare the depending types. I also don't know how to create a circular dependency with the current API.

So, I'm pretty sure we can use the insertion order into the registry from Python as the order of declarations in the output System Verilog. Since Python 3.7, `dict` insertion order is guaranteed to be the iteration order, but I used an OrderedDict to be sure.